src/pages/posts/build-your-own-probot.md

```diff
@@ -0,0 +1,408 @@
+
+Somewhere behind the scenes, Probot has registered the event handlers of your Probot app. A standard app looks like this:
+
+```js
+module.exports = app => {
+  app.on('event', handler)
+}
+```
+
+Where `app` is an instance of the `Application` class, which is (sort of) an extension of `EventEmitter! We'll look at that shortly - there are still a things that Probot does before running your code, but first...
+
+### Why an EventEmitter API?
+
+To me, it's the inherent simplicity of "something happened, so now run this code." When an issue is opened on GitHub, run this code. This let's you focus on your app code, and not care what Probot is doing under the hood. Similar to something like Express's `app.get()`, but this let's you register multiple event handlers for the same event, and have an infinite list of events (since they're just strings).
+
+## Authentication
+
+Next, Probot looks for a particular value in the JSON payload, `installation.id`. Along with your App's credentials (a Private Key and App ID), it uses that installation ID to create a temporary API token.
+
+This was the most complicated part of Probot for a long time, until [@gr2m](https://twitter.com/gr2m) pulled it out into its own repository, [`@octokit/app.js`](https://github.com/octokit/app.js). I'll talk about what it does, then show some code using it!
+
+### Two ways for GitHub Apps to authenticate
+
+If you're here, I'll guess that you're familiar with what GitHub Apps are. They're an "actor" on GitHub, and can authenticate in two distinct ways:
+
+* As itself, as an app
+* As an installation. When you install an app on an account (a user or an organization), that creates an installation. Each installation has a unique `id` (the aforementioned payload value), and we can use that to create an API access token.
+
+Usually, when your Probot app calls `context.github...`, `github` is an authenticated `Octokit` client that uses an **installation token**. We can get one of those by creating an access token as the app and then using that to create an access token for an installation, given its ID.
+
+Authenticating as the app is done by combining the `APP_ID` and `PRIVATE_KEY`, using JWTs:
+
+```js
+const jsonwebtoken = require('jsonwebtoken')
+// A helper function for generated an access token using the App ID and Private Key
+function getSignedJsonWebToken({ app_id, privateKey }) {
+  const now = Math.floor(Date.now() / 1000)
+  const payload = {
+    iat: now, // Issued at time
+    exp: now + 60 * 10 - 30, // JWT expiration time (10 minute maximum, 30 seconds of leeway)
```

Comment:
If you want to grant leeway, you should be adding 30 seconds, not taking it away. 🙂 

```suggestion
    exp: now + 60 * 10 + 30, // JWT expiration time (10 minute maximum, 30 seconds of leeway)
```

Reply:
Hm - you're making sense here, but this is directly copied from Probot. I'm in favor of leaving as-is even if its not correct, maybe with a disclaimer 🤔

Reply:
I think perhaps the problem is just the use of the word "leeway". Looks like the subtraction is intentional per the original issue and PR:

- https://github.com/probot/probot/issues/942
- https://github.com/octokit/app.js/pull/36

Reply:
It's more of a preventative safety net to avoid potentially running into "...the maximum expiration time of 10 minutes, after which the API will start returning a 401 error".

Reply:
I'll just remove the second part of the comment, that level of detail isn't really necessary here 👍

Reply:
https://github.com/octokit/app.js/pull/62

Reply:
Test

Reply:
Another reply which will open the "Leave a review" entry field.

Reply:
One more test. 🤔